### ISSUE TITLE
Equality of different array iterators succeeds

### ISSUE THREAD

**Issue by vkliatchko** [jasmine/jasmine#1121]
The following spec runs without failures:

``` js
"use strict";
describe("test", function () {
    it("compare arrays", function () {
        let a = [1, 2];
        let b = [3, 4];
        expect(a.entries()).toEqual(b.entries());
    });
});
```

Should it not fail?

**Comment by chrisdeely** [comment]
The ```Array.entries()``` method returns an ```Array Iterator``` which doesn't expose any enumerable keys, which are used to evaluate equality in ```toEqual()```. Since the keys for both amount to an empty array, the objects are considered functionally equal.

The quick solution would be to use ```toBe()``` which uses a strict equality comparison, if your code structure would allow for it.

Failing that, a custom matcher could be implemented, or a map within the test.

**Comment by Gerg** [comment]
Jasmine's default equality tester is based off of isEqual from [Underscore.js](http://underscorejs.org/). I just tested it, and Underscore is also not able to handle Array Iterator equality. 

You can define your own [custom equality tester](http://jasmine.github.io/2.5/custom_equality.html)   that accounts for Array Iterators.

I would also be willing to look at a PR that modifies https://github.com/jasmine/jasmine/blob/master/src/core/matchers/matchersUtil.js to make Jasmine support Array Iterators out of the box.

**Comment by vkliatchko** [comment]
Are you suggesting I give you a PR modifying the current behavior of toEnqual for iterators? The current behavior could definitely be surprising and cause bugs. At the same time, the change would not be backward compatible and could break some tests. It also brings a couple of questions:

1. There does not seem to exist an obvious way to match iterators without mutating them. That could be cause bugs in its own way.

2. There does not seem to be a way to detect iterators other than checking for presence of the "next" method. That does not feel reliable.

Thoughts?

**Comment by vkliatchko** [comment]
Sorry. Did not mean to close. Clicked wrong button.

**Comment by zion1129** [comment]
I would be curious to know if this would run without failure if you defined your variables with var instead of let?  We all know let and const don't get hoisted and you would think that code would work as expected but still I would be curious to know...

**Comment by vkliatchko** [comment]
I tried this code with "var". It produces the same behavior, i.e., it reports a success while it would be reasonable to expect a failure. I am not sure why "var" would change this.

Since that there does not seem a solution that overcomes the concerns I pointed out above (not mutating and detecting iterators), I will close the issue out. The work-around I have been using is to wrap the iterators in "Array.from". It might be costly but works.
